language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/ExMemory.hs

```diff
@@ -0,0 +1,118 @@
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE DerivingVia #-}
+{-# LANGUAGE DataKinds #-}
+
+module Language.PlutusCore.Evaluation.Machine.ExMemory
+( Plain
+, WithMemory
+, ExMemory(..)
+, ExCPU(..)
+, GenericExMemoryUsage(..)
+, ExMemoryUsage(..)
+, withMemory
+) where
+
+{- Note [Memory accounting]
+Each operation costs a certain amount of memory. Plutus counts this usage via
+ExMemory units, which correspond to machine words (64bit). First, the cost of
+the initial AST is added to the budget. Then each operation requires a certain
+amount of memory. Builtins may require different amounts of memory depending on
+the input size. Memory cost is only counted on creation of a value, so passing
```

Comment:
This reading seems slightly off to me. I thought we wanted to talk about our memory *residency*, i.e. if things leave the environment then they're not counted any more. This note makes it sound more like CPU, where things are drawn from the budget and never "returned".

Reply:
Also, there's much more we could say here. How do we determine how much memory something uses? We assume it's one word (for the constructor tag?) plus its fields? What about primitives etc.?

Reply:
(As you've probably realised, I *really* like verbose notes about the design decisions in our codebase. They've been immensely useful thus far - let's keep doing them!)